dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1022 @@
+    unnecessary.
+  * For instances, `[Activity`s can be on static or instance methods
+  * Inheritance rules apply from the earlier activity definition section
+  * Users can use the lower-level `AddActivityDefinition` but we left this name unwieldy on purpose, we prefer
+    `AddActivities`
+* Since workflows are not currently implemented, at least one activity is required
+* Activities can be sync or async
+  * For future `Workflow.ExecuteActivity` uses, we have confirmed .NET can disambiguate in an overload between `Task`
+    return type and generic return type with no constraints
+  * Async activities are executed using the normal scheduler
+    * No need yet for a custom scheduler
+  * Sync activities are executed with `ThreadPool`
+    * No need yet for a custom executor
+* As much validation and non-async initialization is done in the constructor as possible
+* While it may be tempting to make `Worker` implement `IAsyncDisposable`, we want users to explicitly handle fatal
+  errors out of `ExecuteAsync`
+  * We can't throw out of `await using` blocks
+* Notice a lack of `StartAsync`, we want users to handle fatal errors from `ExecuteAsync`
+* Notice a lack of `ShutdownAsync`, we want users to use the cancellation token
+* `ExecuteAsync(CancellationToken)` was specifically done this way to match
+  `Microsoft.Extensions.Hosting.BackgroundService.ExecuteAsync`
+
+## Replayer
+
+Looks like:
+
+```csharp
+namespace Temporal.Worker;
+
+class WorkflowReplayer
+{
+    public WorkflowReplayer(TemporalWorkerOptions options) { /*...*/ }
+
+    public async Task<WorkflowReplayResult> ReplayWorkflow(
+        WorkflowHistory history, bool throwOnReplayFailure = true) { /*...*/ }
+
+#if NETCOREAPP3_0_OR_GREATER
+
+    public IAsyncEnumerator<WorkflowReplayResult> ReplayWorkflows(
+      IAsyncEnumerator<WorkflowHistory> histories, bool throwOnReplayFailure = true) { /*...*/ }
+
+#endif
+}
+```
+
+Notes:
+
+* We have intentionally called this `WorkflowReplayer` in .NET since this kind of wordiness is clearer
+* Will need to accept a core runtime in the options here, not sure if we want that as part of worker options or if
+  replayer really needs its own options set that's 99% the same with worker. Maybe just a param.
+
+## Data Conversion
+
+Looks like:
+
+```csharp
+namespace Temporal.Converters;
+
+public record DataConverter(
+    IPayloadConverter PayloadConverter,
+    IFailureConverter FailureConverter,
+    IPayloadCodec? PayloadCodec
+)
+{
+    public static readonly DataConverter Default = new();
+
+    public DataConverter() :
+        this(new PayloadConverter(), new FailureConverter(), null)
+    { }
+}
+
+public interface IPayloadConverter
+{
+    IEnumerable<Temporal.Api.Common.V1.Payload> ToPayloads(
+        IReadOnlyCollection<object> values);
+
+    IEnumerable ToValues(
+        IReadOnlyCollection<Temporal.Api.Common.V1.Payload> payloads, IReadOnlyCollection<Type>? types = null);
```

Comment:
Is the way this works that the two collections must be the same size and the type for each payload is specified by the corresponding index in the types list? Or are there type tags inside the payloads and the types can be in any order?

If the orders must match, this seems easily error prone

Reply:
> If the orders must match, this seems easily error prone

They must match (same size and order) and it is error prone. In Python where we do similar, we disregard if the count isn't the same. I put a lot of thought into this.

* It is an optional argument
* This is a low-level converter implementer concern. The caller is usually me inside the SDK which ensures proper values.
* For those converting manually, they'll be using some higher level helper that will probably only do one value at a time anyways (e.g. `MyConverter.ToValue("foo")`) that then calls this with the generic type there
* We choose at the lowest level to have a collection form of this instead of single payload conversion, but maybe we should change that to just be single? But I'd still provide a `Type` not a generic due to how I have to do it at runtime.
* Technically I could have `ToValues(IReadOnlyCollection<KeyValuePair<Payload, Type>>)` but that assumes I always have a type...maybe I should assume that? In cases where the type is known from the payload, e.g. a proto, the type can be determined without this parameter...but I guess technically they have to provide _some_ type at the call site even if just an object

After some thought, I am considering changing this to:

```
object? ToValue(Payload payload, Type type)
```

With an extension for:

```
T ToValue<T>(Payload payload)
```

This would match https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializer.deserialize. I am trying to think through the ramifications of singular value payload-to-value conversion instead of multiple. I can't think of any (like I can with codec).

Reply:
`ToValues(IReadOnlyCollection<KeyValuePair<Payload, Type>>)` was what I was going to suggest, but is maybe a tiny bit annoying for people to zip together. It is safer, though.

I think only having `ToValue` is also very reasonable. That's how most deserialization API's I've ever seen work. Just more obvious.

Reply:
:+1: I'll change it. I'll also change payload codec to properly be async-capable.

Reply:
I have changed payload converter to only deal with single values, changed payload codec to support async, and added extension methods to these converters.

(leaving this discussion unresolved for posterity)